### PR TITLE
Implement WoodokuBoard and Significant refactoring 

### LINKED ISSUE #4: Finish implementing class WoodokuBoard

### PR DESCRIPTION
**Pull request by Xiaorui-Huang**
### Significant refactoring

1. docstring update (style and content)

	- to google style
	- terminology update according to sudoku [Wikipedia](https://en.wikipedia.org/wiki/Glossary_of_Sudoku#Terminology_and_grid_layout) page
2. revised some import statement
3. updated config for _.gitignore_ and _Pipfile_ from virtual environment manager _pipenv_
4. updated some public api to private. new public api. refactored private api.
5. added exception folder, all future exceptions in Woodoku should perhaps go in the same file

resolves #4

### CODE REVIEW

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 1]
```diff
@@ -22,86 +25,167 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
+            bool: if the `shape` fits the board
         """
-        Check if the woodoku shape can fit into the board. If all current shapes are not be able to add, the game fails.
-        :return: if can fit.
-        """
-        pass
+        # TODO: improve performance
+        rng = np.random.default_rng()
+        all_blocks = [(i, j) for i in range(9) for j in range(9)]
+        rng.shuffle(all_blocks)
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py:90]: Sure, ok, fine.

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 2]
```diff
@@ -22,86 +25,167 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
+            CannotPlaceShapeError: raised when checking a invalid block
 
-        :param blocks_coord:
-        :return:
+        Returns:
+            bool: if any coordinates in `blocks_coord` is occupied
         """
-        pass
+        # TODO: raise the Error, according to docstring, the code below is just
+        # a placeholder
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]: Might want to delete this comment. Instead, create a new issue with detailed explanation.

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 3]
```diff
@@ -22,86 +25,167 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
+            shape (WoodokuShape): a woodoku shape for validation
+
+        Returns:
+            bool: if the `shape` fits the board
         """
-        Check if the woodoku shape can fit into the board. If all current shapes are not be able to add, the game fails.
-        :return: if can fit.
-        """
-        pass
+        # TODO: improve performance
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]: Might want to delete this comment. Instead, create a new issue with detailed explanation.

TBH, performance on the game is not the highest priority.

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 4]
```diff
@@ -22,86 +25,167 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
+        rng = np.random.default_rng()
+        all_blocks = [(i, j) for i in range(9) for j in range(9)]
+        rng.shuffle(all_blocks)
+        print(all_blocks)
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion

```

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 5]
```diff
@@ -22,86 +25,167 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
         """
-        Check if the woodoku shape can be placed into the board at location (x, y) coordinates as the left upper corner
-        of the shape.
-        :param shape: the shape
-        :param x: x coordinate
-        :param y: y coordinate
-        :return: if can be placed at location.
-        """
-        pass
+        blocks = self.__map_shape_to_board_at(shape, x, y)
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]: Is the signature of this function defined? 

Plus, maybe only one of `WoodokuBoard.__map_shape_to_board_at` and `WoodokuShape.map_to_board_at` is needed.

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 6]
```diff
@@ -22,86 +25,167 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
+
+        Args:
+            blocks_coord (List[Tuple[int, int]]): list of coordinates to check
+
+        Raises:
+            CannotPlaceShapeError: raised when checking a invalid block
 
-        :param blocks_coord:
-        :return:
+        Returns:
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py:45]: The return value comment contradicts with the method docstring on where to return if **ANY** or **ALL** blocks are occupied.

**Comment by Xiaorui-Huang** [comment]
rebase onto current develop branch, will implement woodoku board __str__()

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 7]
```diff
@@ -25,89 +28,171 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
-        Check if the woodoku shape can be placed into the board at location (x, y) coordinates as the left upper corner
-        of the shape.
-        :param shape: the shape
-        :param x: x coordinate
-        :param y: y coordinate
-        :return: if can be placed at location.
-        """
-        pass
+        blocks = self.__map_shape_to_board_at(shape, x, y)
+        return not self.__representation.is_occupied(blocks)
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]: The definition for `_WoodokuBoardRepresentation.is_occupied` is unclear. If it returns true when **ALL** blocks are occpied, then the code here is incorrect.

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 8]
```diff
@@ -25,89 +28,171 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
-        :return: All the coords in this 3x3 block.
+        info = {"row": 0, "column": 0, "box": 0}
+        rep = self.__representation
+        group_blocks = set()  # use set to handle overlapping group removal
+        for index in range(9):
+            row_blocks = self.__get_row_coords(index)
+            col_blocks = self.__get_col_coords(index)
+            box_blocks = self.__get_box_coords(index)
+
+            if rep.is_occupied(row_blocks):
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py:162]: The definition for `_WoodokuBoardRepresentation.is_occupied` is unclear. If it returns true when **ANY** blocks are occpied, then the code here is incorrect.

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 9]
```diff
@@ -25,89 +28,171 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
+        #  3 -> 1,0
+        #  4 -> 1,1
+        #  5 -> 1,2
+        #  6 -> 2,0
+        #  7 -> 2,1
+        #  8 -> 2,2
+        x, y = index // 3, index % 3
+
+        # map to 9x9 coordinates
+        x, y = x * 3, y * 3
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py:210]: Smart

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 10]
```diff
@@ -25,89 +28,171 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
-    def __find_winning_blocks(self) -> Dict[str, int]:
-        """
-        Check current board and see if there is any winning blocks such as complete rows, columns or 3x3 block.
-        :return: A dictionary mapping from name of the winning block to the number of winning block.
-            e.g. { "row": 1, "column": 2, "3x3 block": 0 }
-        """
-        pass
+    def __find_groups(self) -> Tuple[Dict[str, int], Set[Tuple[int, int]]]:
+        """Check current board and see if there is any groups such as
+        complete rows, columns or 3x3 box and report them. Remove all groups.
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]: Seems that it is not removing all the groups. Might want to update the docstring.

File: Woodoku/src/Exceptions/Exceptions.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+from mimetypes import init
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Exceptions/Exceptions.py]: What is the use of this package?

**Review by EdwardHaoranLee — DISMISSED** [review]

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 12]
```diff
@@ -43,6 +43,10 @@ def is_occupied(self, blocks_coord: List[Tuple[int, int]]) -> bool:
         # TODO: raise the Error, according to docstring, the code below is just
         # a placeholder
         raise CannotPlaceShapeError(0, 0)
+    
+    # TODO: implement str representation
+    def __str__(self) -> str:
+        raise NotImplementedError()
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py:49]: @YANLING-H Might need to also take this into consideration

File: Woodoku/src/Exceptions/Exceptions.py (hunk 11 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/Exceptions/Exceptions.py]: Hahaha, intellisense's leftover wreck 😅
Nice catch

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 5 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/Entity/WoodokuBoard.py]: This line was a mistake from a previous version of the function signature, all should be `WoodokuShape.map_to_board_at`

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 6 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/Entity/WoodokuBoard.py:45]: Will update it `is_occupied` and `is_not_occupied` both of which checks **all** blocks

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 7 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/Entity/WoodokuBoard.py]: Will request and use`_WoodokuBoardRepresentation.is_occupied` checking for **all** blocks to resolve the problem.

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 8 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/Entity/WoodokuBoard.py:162]: Will request and use`_WoodokuBoardRepresentation.is_not_occupied` which checks for **all** blocks to resolve the problem

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 13]
```diff
@@ -25,89 +25,191 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
-        :param x: x coordinate
-        :param y: y coordinate
-        :return: if can be placed at location.
-        """
-        pass
+        blocks = shape.map_to_board_at(x,y)
+        try:
+            return self.__representation.is_not_occupied(blocks)
+        except ShapeOutOfBoardError:
+            return False
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]: The upper level game logic might still need a difference between that a shape cannot be added at location and the coordinates were incorrect. Is there any particular reason why we are not throwing this error?

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 14]
```diff
@@ -25,89 +25,191 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
+        #  7 -> 2,1
+        #  8 -> 2,2
+        x, y = index // 3, index % 3
+
+        # map to 9x9 coordinates
+        x, y = x * 3, y * 3
+        return [(row, col) for row in range(x, x + 3) for col in range(y, y + 3)]
+    
+    def __str__(self) -> str:
+        return str(self.__representation)
```
Comment (EdwardHaoranLee) [inline Woodoku/src/Entity/WoodokuBoard.py]: It should include score as part of `__str__` print out. Mainly use this method in the `CommandLineUI`.

But could leave this method blank for now.

**Review by EdwardHaoranLee — CHANGES_REQUESTED** [review]

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 14 above)
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion
		#TODO: include score in print out, should be implemented along with CommandLineUI
        pass
```

File: Woodoku/src/Entity/WoodokuBoard.py (hunk 13 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/Entity/WoodokuBoard.py]: I agree, the only scenario for handling this error that I can think of is for printing the error to `CommandLineUI` . For now, I'll let it throw

Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion
            return self.__representation.is_not_occupied(blocks)
```

File: Woodoku/src/Entity/WoodokuBoard.py [hunk 15]
```diff
@@ -25,89 +25,191 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
...
 
     def can_add_shape_at_location(self, shape: WoodokuShape, x: int, y: int) -> bool:
+        """Check if the woodoku `shape` can be placed into the board at location
+        `(x, y)` coordinates as the left upper corner of the shape.
+
+        Args:
+            shape (WoodokuShape): The shape needed to be checked
+            x (int): x coordinate
+            y (int): y coordinate
+
```
Comment (Xiaorui-Huang) [inline Woodoku/src/Entity/WoodokuBoard.py]:
```suggestion
            y (int): y coordinate
        Raises: 
            ShapeOutOfBoundError: if any block in `blocks` is invalid

```

**Review by EdwardHaoranLee — APPROVED** [review]

**Pull request merged**